depthai_ros_driver/include/depthai_ros_driver/device_ros.hpp

```diff
@@ -85,18 +85,27 @@ class DeviceROS : public dai::DeviceBase {
 protected:
     // create stream based on name at the time of subscription
     template <class MsgType>
-    auto generate_cb_lambda(std::string name) {
-        auto conn = this->getConnection();
-        const auto core_sub_lambda = [&, stream = dai::XLinkStream{*conn, name, dai::XLINK_USB_BUFFER_MAX_SIZE}](
-                                             MsgType& msg, dai::RawBuffer& data) {
+    auto generate_cb_lambda(std::unique_ptr<dai::XLinkStream>& stream) -> boost::function<void(const boost::shared_ptr<MsgType const >&)> {
+        // auto conn = this->getConnection();
+        // const auto core_sub_lambda = [stream = dai::XLinkStream{*conn, name, dai::XLINK_USB_BUFFER_MAX_SIZE}](
+        const auto core_sub_lambda = [&stream] (
+                                             const boost::shared_ptr<MsgType const >& msg) {
             Guard guard([] { ROS_ERROR("Communication failed: Device error or misconfiguration."); });
 
-            // TODO:  convert msg to data here somehow
-            std::vector<std::uint8_t> serialized = dai::StreamPacketParser::serializeMessage(data);
-            stream.write(serialized);
+            // convert msg to data
+            std::stringstream buffer;
+            msgpack::pack(buffer, *msg);
+            buffer.seekg(0);
+
+            const std::string str(buffer.str());
+            const std::vector<std::uint8_t> serialized(str.cbegin(), str.cend()); // copying?
+            stream->write(serialized);
```

Comment:
Let me create a PR for a begin and end (so we don't need to create buffer or str or serialized

Reply:
They already have write with uint8_t (as well as void) pointer and size. That could be used to save at least 1 copy. Using https://github.com/msgpack/msgpack-c/wiki/v2_0_cpp_packer#sbuffer might also make sense